skfuzzy/defuzzify/defuzz.py

```diff
@@ -58,7 +58,43 @@ def centroid(x, mfx):
     skfuzzy.defuzzify.defuzz, skfuzzy.defuzzify.dcentroid
 
     """
-    return (x * mfx).sum() / np.fmax(mfx.sum(),
+
+    '''
+    As we suppose linearity between each pair of points of x, we can calculate the exact area of the figure
+    (a triangle or a rectangle).
+    '''
+    sum_moment_area = 0.0
+    sum_area = 0.0
+    #If the membership function is a singleton fuzzy set:
+    if len(x)==1:
+        return x[0]*mfx[0]/np.fmax(mfx[0],
+                                     np.finfo(float).eps).astype(float)
+    #else return the sum of moment*area/sum of area
+    for i in range(1,len(x)):
```

Comment:
For this PR, keep the for loops. Reason being that this is essentially exactly the formulation Cython would need. As a quick follow up, we can then use this PR as a base to benchmark alternative approaches.